src/OpenTelemetry.Instrumentation.AspNetCore/TracerProviderBuilderExtensions.cs

```diff
@@ -70,17 +70,24 @@ public static TracerProviderBuilder AddAspNetCoreInstrumentation(this TracerProv
                 builder.ConfigureServices(services => services.Configure(name, configureAspNetCoreInstrumentationOptions));
             }
 
-            return builder.ConfigureBuilder((sp, builder) =>
+            if (builder is IDeferredTracerProviderBuilder deferredTracerProviderBuilder)
+            {
+                deferredTracerProviderBuilder.Configure((sp, builder) =>
+                {
+                    AddAspNetCoreInstrumentationSources(builder, sp);
```

Comment:
Would it be weird or not possible to do this within builder.AddInstrumentation...

```csharp
            return builder.AddInstrumentation(sp =>
            {
                var options = sp.GetRequiredService<IOptionsMonitor<AspNetCoreInstrumentationOptions>>().Get(name);

                // I'm guessing it's too late or could cause problems to futz with the builder here.
                AddAspNetCoreInstrumentationSources(builder, sp);

                return BuildAspNetCoreInstrumentation(options);
            });
```

Reply:
I think the only reason it is getting weird is because the ActivitySource is being resolved from the service provider to read it's name. We've talked with Fowler in the past and it was a mistake for the ActivitySource to be rooted in the container. There was no intention for that to be a user-replacable service. For compat we probably won't change it, but we also wouldn't suggest anyone to replace it. I'd suggest hard-coding a name for the source, the same as we do for .NET 6.

Reply:
If you do that you will get a `NotSupportedException` with the message: `Builder cannot be configured during TracerProvider construction.`

The reason for that is the "builder" that is coming into this method is just a wrapper over `IServiceCollection`. Anything you do to it just dumps a delegate into `IServiceCollection` to be executed later once the `IServiceProvider` is available. What this code is doing is accessing the "builder" `IServiceCollection` wrapper via a closure. At the point the code runs `IServiceProvider` has already been created so anything performed on `IServiceCollection` is ignored. `IServiceCollection` doesn't throw for that it just silently fails but I added the `NotSupportedException` on the builder to help users catch the mistake.

Reply:
To @noahfalk's point, yes, this case is unusual due to what Asp.Net Core is doing with their source. We could add something like `AddSource(Func<IServiceProvider, string[]> implementationFactory)` and then we could clean this up but there probably isn't a strong need for that.

Reply:
My suggestion (not that it needs to be in this PR) is to replace lines 108-121 below with this:

```C#
#if NET7_0_OR_GREATER
            builder.AddSource(HttpInListener.AspNetCoreActivitySourceName);
#else
```

ActivitySource names (similar to Meter, DiagnosticSource, Logger, and EventSource) are intended  to be a fixed part of the contract a telemetry data provider is making with consumers. If the producer ever changes that name it is a breaking change. This means you shouldn't need to dynamically compute the string from the ServiceProvider, so in turn you won't need ConfigureBuilder() or a new AddSource() overload.

Reply:
Opened https://github.com/open-telemetry/opentelemetry-dotnet/issues/4124